libs/core/executors/include/hpx/executors/fork_join_executor.hpp

```diff
@@ -428,28 +450,27 @@ namespace hpx { namespace execution { namespace experimental {
                 // Set the data for this parallel region
                 element_function_ = static_cast<void*>(&f);
                 shape_ = static_cast<void const*>(&shape);
-                size_ = hpx::util::size(shape);
-                auto argument_pack = hpx::make_tuple<>(HPX_FORWARD(Ts, ts)...);
+                auto argument_pack =
+                    hpx::forward_as_tuple(HPX_FORWARD(Ts, ts)...);
                 argument_pack_ = static_cast<void*>(&argument_pack);
-                thread_function_helper_ =
-                    static_cast<thread_function_helper_type*>(
-                        &thread_function_helper<typename std::decay<F>::type,
-                            typename std::decay<S>::type,
-                            decltype(argument_pack)>::call);
+                thread_function_helper_ = &thread_function_helper<F, S,
+                    decltype(argument_pack)>::call;
 
                 // Signal all worker threads to start partitioning work for
                 // themselves, and then starting the actual work.
                 set_state_all(thread_state::partitioning_work);
 
                 // Start work on the main thread.
-                thread_function_helper<typename std::decay<F>::type,
-                    typename std::decay<S>::type,
-                    decltype(argument_pack)>::call(element_function_, shape_,
-                    argument_pack_, main_thread_, num_threads_, schedule_,
-                    queues_, thread_states_, exception_mutex_, exception_);
+                thread_function_helper<F, S, decltype(argument_pack)>::call(
+                    element_function_, shape_, argument_pack_, main_thread_,
+                    num_threads_, schedule_, queues_, thread_states_,
+                    exception_mutex_, exception_);
 
+                // Wait for all threads to finish their work assigned to
+                // them in this parallel region.
                 wait_state_all(thread_state::idle);
 
+                std::lock_guard l(exception_mutex_);
```

Comment:
I think this lock should not be needed _if_ the states are read/written with acquire/release instead of relaxed.

Reply:
Yes, I understand. Except if it's not ;-) (assuming that the executor is always used from the same main thread, and that it is never invoked concurrently...) Does it hurt to have the lock here?

Reply:
Yeah, I realize you changed some of them to `relaxed` :P 

> Does it hurt to have the lock here?

Would it be nice not to have it there? Yes. Will I (or anyone else) notice that the lock is there? I doubt it ;)

The question is what we gain with the `relaxed` vs what we lose with the lock here.

Reply:
The `relaxed` is still an atomic operation; `release`/`acquire` just additionally prevent operation reordering across it. Thus, we most likely do not gain anything (measurable) from using `relaxed`. The lock will also most likely not be contended, so all we do is adding another two atomic load/store operations.

The `fork_join_executor` is quite sophisticated and not easy to understand. Not having the lock might make understanding it even more difficult.

Reply:
I'm saying I don't mind if the lock is there so if you prefer to have it there let's have it there.

But outside of that I doubt the lock helps in understanding the executor ;) And if `relaxed` doesn't help then why change it?